lib/types.js

```diff
@@ -1251,13 +1266,31 @@ class UnwrappedUnionType extends UnionType {
 
     this._dynamicBranches = null;
     this._bucketIndices = {};
+
+    this.projectionFunction = (val) => this._bucketIndices[getValueBucket(val)];
```

Comment:
Let's consolidate logic to avoid unnecessary fields (e.g. `_bucketIndices` when a projection is present). For example something like:

```javascript
constructor (schema, opts, proj) {
  super(schema, opts);
  if (proj == null && typeof opts?.wrapUnions == 'function') {
    proj = opts.wrapUnions(this.types);
  }
  this._getIndex = proj ? projectionIndexer(proj) : defaultIndexer(this.types);
}
```

`defaultIndexer` encapsulates the current logic, moving it out of the constructor along with `_getBranchIndex`. `projectionIndexer` is lightweight, just adding error handling to the projection.

Reply:
changes in 510d272 - not sure I prefer all of it, as `projectionFn` is now a public interface to `UnwrappedUnionType`. We could possibly set it as a member on the instance of `UnionType`, less transparent but won't become a public interface.

When I tried to make `this._bucketIndices` local variables I noticed that this member is still used by `_copy` - I am unsure whether changes are needed in that function with the projection change?

Reply:
> When I tried to make `this._bucketIndices` local variables I noticed that this member is still used by `_copy` - I am unsure whether changes are needed in that function with the projection change?

changed in bd2c6dd